File: WinForms/CS/HtmlMainDemo/Modules/TileView/TileViewAutoHeightModule.cs
using System.Drawing;
using System.Windows.Forms;
using DevExpress.DXperience.Demos;
using DevExpress.HTML.Demos.Helpers.Tasks;
using DevExpress.Utils.Menu;
using DevExpress.XtraGrid.Views.Tile;

namespace DevExpress.HTML.Demos {
    public partial class TileViewAutoHeightModule : TutorialControlBase {
        public TileViewAutoHeightModule() {
            InitializeComponent();
            if(MainFormHelper.TakeScreens)
                tileView1.AnimateArrival = false;
            tileViewTemplateCodeViewer1.Fill(tileView1);
            gridControl1.DataSource = TaskGenerator.Default.GenerateSource(37);
            tileView1.FocusedRowHandle = 2;

            alertControl1.BeforeFormShow += (s, e) => { e.Location = gridControl1.PointToScreen(Point.Empty); };
            alertControl1.GetDesiredAlertFormWidth += (s, e) => { e.Width = gridControl1.Width; };
            tileView1.KeyDown += TileView1_KeyDown;

            copyMenu = new DXPopupMenu();
            copyMenu.Items.Add(new DXMenuItem("Copy", CopySelected, svgImageCollection1[0], DXMenuItemPriority.Normal));
            tileView1.HtmlElementMouseClick += TileView1_HtmlElementMouseClick;
        }
        DXPopupMenu copyMenu;

        void TileView1_KeyDown(object sender, KeyEventArgs e) {
            if(Control.ModifierKeys.HasFlag(Keys.Control) && e.KeyCode == Keys.C)
                CopySelected(sender, e);
        }
        void TileView1_HtmlElementMouseClick(object sender, TileViewHtmlElementMouseEventArgs e) {
            if(e.MouseArgs.Button == MouseButtons.Right && e.HitInfo.InSelection)
                MenuManagerHelper.ShowMenu(copyMenu, LookAndFeel, MenuManagerHelper.GetMenuManager(LookAndFeel, Parent), this, e.MouseArgs.Location);
        }
        void CopySelected(object sender, System.EventArgs e) {
            var text = tileView1.SelectedHtmlText;
            if(!string.IsNullOrEmpty(text)) {
                DevExpress.Data.Utils.SafeClipboardWin.Instance.SetText(text);
                alertControl1.Show(FindForm(), string.Empty, "Selected text copied to clipboard");
            }
        }
    }
}

File: WinForms/VB/HtmlMainDemo/Modules/TileView/TileViewAutoHeightModule.vb
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.DXperience.Demos
Imports DevExpress.HTML.Demos.Helpers.Tasks
Imports DevExpress.Utils.Menu
Imports DevExpress.XtraGrid.Views.Tile

Namespace DevExpress.HTML.Demos

    Public Partial Class TileViewAutoHeightModule
        Inherits TutorialControlBase

        Public Sub New()
            InitializeComponent()
            If MainFormHelper.TakeScreens Then tileView1.AnimateArrival = False
            tileViewTemplateCodeViewer1.Fill(tileView1)
            gridControl1.DataSource = TaskGenerator.Default.GenerateSource(37)
            tileView1.FocusedRowHandle = 2
            AddHandler alertControl1.BeforeFormShow, Sub(s, e) e.Location = gridControl1.PointToScreen(Point.Empty)
            AddHandler alertControl1.GetDesiredAlertFormWidth, Sub(s, e) e.Width = gridControl1.Width
            AddHandler tileView1.KeyDown, AddressOf TileView1_KeyDown
            copyMenu = New DXPopupMenu()
            copyMenu.Items.Add(New DXMenuItem("Copy", AddressOf CopySelected, svgImageCollection1(0), DXMenuItemPriority.Normal))
            AddHandler tileView1.HtmlElementMouseClick, AddressOf TileView1_HtmlElementMouseClick
        End Sub

        Private copyMenu As DXPopupMenu

        Private Sub TileView1_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
            If ModifierKeys.HasFlag(Keys.Control) AndAlso e.KeyCode = Keys.C Then CopySelected(sender, e)
        End Sub

        Private Sub TileView1_HtmlElementMouseClick(ByVal sender As Object, ByVal e As TileViewHtmlElementMouseEventArgs)
            If e.MouseArgs.Button = MouseButtons.Right AndAlso e.HitInfo.InSelection Then MenuManagerHelper.ShowMenu(copyMenu, LookAndFeel, MenuManagerHelper.GetMenuManager(LookAndFeel, Parent), Me, e.MouseArgs.Location)
        End Sub

        Private Sub CopySelected(ByVal sender As Object, ByVal e As System.EventArgs)
            Dim text = tileView1.SelectedHtmlText
            If Not String.IsNullOrEmpty(text) Then
                Data.Utils.SafeClipboardWin.Instance.SetText(text)
                alertControl1.Show(FindForm(), String.Empty, "Selected text copied to clipboard")
            End If
        End Sub
    End Class
End Namespace
